src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                            // foreach(...)
+                            variable = SyntaxFactory.IdentifierName(variableDeclarator.Identifier);
+                            documentUpdateMethod =
+                                s => new DocumentUpdateInfo(
+                                    parentStatement,
+                                    new[] { parentStatement.ReplaceNode(invocationExpression, initializer.WithAdditionalAnnotations(Simplifier.Annotation)), s });
+                            return true;
+                        }
+                        break;
+                    case SyntaxKind.SimpleAssignmentExpression:
+                        // before a = (from a in b select a).ToList();
+                        // after a = new List<T>();
+                        // foreach(...)
+                        variable = ((AssignmentExpressionSyntax)invocationParent).Left;
+                        documentUpdateMethod =
+                            s => new DocumentUpdateInfo(
+                                parentStatement,
+                                new[] { parentStatement.ReplaceNode(invocationExpression, initializer.WithAdditionalAnnotations(Simplifier.Annotation)), s });
+                        return true;
+                    case SyntaxKind.ReturnStatement:
+                        // before return (from a in b select a).ToList();
+                        // after var list = new List<T>();
+                        // foreach(...)
+                        // return list;
+                        string freeVariableName = GetFreeSymbolName(variableName, source.GetLocation().SourceSpan.Start);
+                        var identifierName = SyntaxFactory.IdentifierName(freeVariableName);
+                        var variableDeclaration =
+                            SyntaxFactory.LocalDeclarationStatement(
+                                SyntaxFactory.VariableDeclaration(
+                                    VarNameIdentifier,
+                                    SyntaxFactory.SingletonSeparatedList(
+                                        SyntaxFactory.VariableDeclarator(
+                                            identifierName.Identifier,
+                                            argumentList: null,
+                                            SyntaxFactory.EqualsValueClause(initializer)))));
+
+                        variable = identifierName;
+                        documentUpdateMethod =
+                            s => new DocumentUpdateInfo(
+                                parentStatement,
+                                new[] { variableDeclaration.WithAdditionalAnnotations(Simplifier.Annotation), s, SyntaxFactory.ReturnStatement(identifierName).WithAdditionalAnnotations(Simplifier.Annotation) });
+                        return true;
+                        // SyntaxKind.ArrowExpressionClause is not supported
+                }
+
+                documentUpdateMethod = default;
```

Comment:
>documentUpdateMethod = default; [](start = 16, length = 31)

I am not sure why are we giving up that easy. Why not inject the loop before the containing statement and replace invocation with the local?

Reply:
> I am not sure why are we giving up that easy. Why not inject the loop before the containing statement and replace invocation with the local? 

It looks like this got marked a s resolved without any explanation and I feels like the transformation that we do for the SyntaxKind.Argument case can be done for any other case.


---
In reply to: [176929670](https://github.com/dotnet/roslyn/pull/25362#discussion_r176929670) [](ancestors = 176929670)

Reply:
This one is covered with TryReplaceWithLocalFunction which is the default behavior. If we give up here, we still handle this with the local function

---
In reply to: [178403246](https://github.com/dotnet/roslyn/pull/25362#discussion_r178403246) [](ancestors = 178403246,176929670)

Reply:
> This one is covered with TryReplaceWithLocalFunction which is the default behavior. If we give up here, we still handle this with the local function

Yes, but it looks like an inline transform might just work. Why the argument case is different?

---
In reply to: [178408593](https://github.com/dotnet/roslyn/pull/25362#discussion_r178408593) [](ancestors = 178408593,178403246,176929670)